base/inference.jl

```diff
@@ -1433,6 +1499,17 @@ function abstract_call(f::ANY, fargs, argtypes::Vector{Any}, vtypes::VarTable, s
 
     if istopfunction(tm, f, :promote_type) || istopfunction(tm, f, :typejoin)
         return Type
+    elseif length(argtypes) == 2 && istopfunction(tm, f, :typename)
+        t = argtypes[2]
+        if isa(t, Const) || isType(t)
+            val = isa(t, Const) ? t.val : t.parameters[1]
```

Comment:
this might encounter `Vararg` or `TypeVar`, and will need to handle those cases also